src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -497,80 +154,72 @@ class LedgerImpl(
+        case Some(currentTd) =>
+          if (isPossibleNewBestBlock(block.header, currentBestBlock.header)) {
+            blockImport.importToTop(block, currentBestBlock, currentTd)(blockExecutionContext)
+          } else {
+            blockImport.reorganise(block, currentBestBlock, currentTd)
+          }
+
+        case None =>
+          Future.successful(BlockImportFailed(s"Couldn't get total difficulty for current best block with hash: $hash"))
+
+      }
     }
   }
 
-  def validateBlockBeforeExecution(block: Block): Either[ValidationBeforeExecError, BlockExecutionSuccess] = {
-    consensus.validators.validateBlockBeforeExecution(
-      block = block,
-      getBlockHeaderByHash = getBlockHeaderFromChainOrQueue,
-      getNBlocksBack = getNBlocksBackFromChainOrQueue
-    )
+  private def isBlockADuplicate(block: BlockHeader, currentBestBlockNumber: BigInt): Boolean = {
+    val hash = block.hash
+    blockchain.getBlockByHash(hash).isDefined && block.number <= currentBestBlockNumber || blockQueue.isQueued(hash)
   }
 
-  private[ledger] def validateBlockAfterExecution(
-    block: Block,
-    stateRootHash: ByteString,
-    receipts: Seq[Receipt],
-    gasUsed: BigInt
-  ): Either[BlockExecutionError, BlockExecutionSuccess] = {
-
-    consensus.validators.validateBlockAfterExecution(
-      block = block,
-      stateRootHash = stateRootHash,
-      receipts = receipts,
-      gasUsed = gasUsed
-    )
-  }
+  private def isPossibleNewBestBlock(newBlock: BlockHeader, currentBestBlock: BlockHeader): Boolean =
+    newBlock.parentHash == currentBestBlock.hash && newBlock.number == currentBestBlock.number + 1
 
-  /**
-    * This function updates worldState transferring balance from drainList accounts to refundContract address
-    *
-    * @param worldState Initial world state
-    * @param daoForkConfig Dao fork configuration with drainList and refundContract config
-    * @return Updated world state proxy
-    */
-  private def drainDaoForkAccounts(worldState: InMemoryWorldStateProxy, daoForkConfig: DaoForkConfig): InMemoryWorldStateProxy = {
-
-    daoForkConfig.refundContract match {
-      case Some(refundContractAddress) =>
-        daoForkConfig.drainList.foldLeft(worldState) { (ws, address) =>
-          ws.getAccount(address)
-            .map(account => ws.transfer(from = address, to = refundContractAddress, account.balance))
-            .getOrElse(ws)
-        }
-      case None => worldState
+  override def resolveBranch(headers: Seq[BlockHeader]): BranchResolutionResult = {
+    if (!branchResolution.areHeadersFormChain(headers) || headers.last.number < blockchain.getBestBlockNumber()) {
+      InvalidBranch
+    } else {
+      // Dealing with a situation when genesis block is included in the received headers,
+      // which may happen in the early block of private networks
+      val result = for {
+        genesisHeader      <- blockchain.getBlockHeaderByNumber(0)
```

Comment:
actually blockchain has a method called `genesisHeader`. 

Also i am wondering why not to move this method to`BranchResolution`, and let ledeger only delegete to it in `resolveBranch` ?  As now situation is kinda weird, we have bunch of methods related to resolution in one module, and main resolution method in other module using those methods.